dask/array/core.py

```diff
@@ -880,31 +884,85 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
         raise ValueError("Different number of sources [%d] and targets [%d] than regions [%d]"
                          % (len(sources), len(targets), len(regions)))
 
-    updates = {}
-    keys = []
+    # Optimize all sources together
+    sources_dsk = sharedict.merge(*[e.__dask_graph__() for e in sources])
+    sources_dsk = Array.__dask_optimize__(
+        sources_dsk,
+        [e.__dask_keys__() for e in sources]
+    )
+
+    tgt_dsks = []
+    store_keys = []
+    store_dsks = []
+    if keep:
+        load_names = []
+        load_dsks = []
     for tgt, src, reg in zip(targets, sources, regions):
         # if out is a delayed object update dictionary accordingly
         try:
-            dsk = {}
-            dsk.update(tgt.dask)
+            each_tgt_dsk = {}
+            each_tgt_dsk.update(tgt.dask)
             tgt = tgt.key
         except AttributeError:
-            dsk = {}
+            each_tgt_dsk = {}
 
-        update = insert_to_ooc(src, tgt, lock=lock, region=reg)
-        keys.extend(update)
+        src = Array(
+            sources_dsk,
+            src.name,
+            src.chunks,
+            src.dtype
+        )
 
-        update.update(dsk)
-        updates.update(update)
+        each_store_dsk = insert_to_ooc(
+            src, tgt, lock=lock, region=reg, keep=keep
+        )
 
-    name = 'store-' + tokenize(*keys)
-    dsk = sharedict.merge((name, updates), *[src.dask for src in sources])
-    if compute:
-        compute_as_if_collection(Array, dsk, keys, **kwargs)
+        if keep:
+            load_names.append('load-store-%s' % src.name)
+            load_dsks.append(retrieve_from_ooc(
+                each_store_dsk.keys(),
+                each_store_dsk
+            ))
+
+        tgt_dsks.append(each_tgt_dsk)
+
+        store_keys.extend(each_store_dsk.keys())
+        store_dsks.append(each_store_dsk)
+
+    store_dsks_mrg = sharedict.merge(*concatv(
+        store_dsks, tgt_dsks, [sources_dsk]
+    ))
+
+    result = None
+    if keep:
```

Comment:
I'd personally change this logic to:

```python
if keep:
    if compute:
        ...
    return tuple(result)
elif compute:
    dyld.compute()
    return None
else:
    return dyld
```

Reduces nesting, and doesn't set `result` to `None` far from where it's used. Again, just a personal preference, no strong thoughts.

Reply:
Ok, I'll give it a look.

Reply:
Maybe I'm not understanding the concern here. The two `compute` cases now share a fair bit of code. So breaking them apart will create some redundancy.

Is the concern more that the `return` statement has an unclear value? If so, we could add `return` statements to the respective branches instead. Would that help?

Reply:
> Is the concern more that the return statement has an unclear value? If so, we could add return statements to the respective branches instead. Would that help?

That's what I was suggesting, the dedent of the non-keep logic required a bit of duplication which I hadn't thought about. To be perfectly clear, here's my suggestion:

```python
if keep:
    if compute:
	store_dlyds = [Delayed(k, store_dsks_mrg) for k in store_keys]
	store_dlyds = persist(*store_dlyds)
	store_dsks_mrg = sharedict.merge(*[e.dask for e in store_dlyds])

    load_dsks_mrg = sharedict.merge(store_dsks_mrg, *load_dsks)

    result = []
    for each_load_name in load_names:
	result.append(Array(
	    load_dsks_mrg,
	    each_load_name,
	    src.chunks,
	    src.dtype
	))
    return tuple(result)
else:
    name = 'store-' + tokenize(*store_keys)
    dsk = sharedict.merge({name: store_keys}, store_dsks_mrg)
    dlyd = Delayed(name, dsk)

    if compute:
	dlyd.compute()
        return None
    else:
	return dlyd
```

Again, not a huge deal, just a suggestion.

Reply:
Thanks. That's clear. Have updated.